README.md
# virtual-inf
静岡大学 バーチャル浜松キャンパス  
Shizuoka University Virtual Hamamatsu Campus

静岡大学の浜松キャンパスをバーチャルリアリティで楽しむことができます。  
[静岡大学情報学部 ONLINE OPEN CAMPUS 2020](https://www.inf.shizuoka.ac.jp/opencampus2020/index.html)のコンテンツとして制作したものですが、オープンキャンパス終了後も開発および公開を続けております。

## Getting Started

https://xr-hamaria.github.io/virtual-inf にて公開しています。

一般的なウェブブラウザ・ウェブサーバであれば問題なく動作します。  
モデルデータの容量が30MB程度となります。十分な通信速度を確保してください。

左上の「VRモード」を選択すると、一人称視点でキャンパスを移動することができます。  
PCの場合はWASDキー/矢印キーで、スマホやタブレットの場合はタップで操作が可能です。

一部のVR対応機器ならびに対応ブラウザでページを開くと、画面下部に「ENTER VR」と表示されます。  
こちらを選択するとHMDでのVR体験ができます。Oculus Questで動作確認済みです。  
SteamVRにも対応しました。Mozilla Firefoxでページを起動し、VRへのアクセスを許可して、画面下部の「ENTER VR」をクリックしてください。

[ローカルテストサーバーで実行する](https://github.com/xr-hamaria/virtual-inf/wiki/ローカルテストサーバーで実行する)

## License

[MIT License](https://github.com/xr-hamaria/virtual-inf/blob/master/LICENSE)  
ただし、学章等のデータの利用については[静岡大学学章等に関する規則](http://reiki.adb.shizuoka.ac.jp/act/frame/frame110000075.htm)をご参照ください。  
また、キャンパスデータには一部OpenStreetMapのデータを含みますので、[https://www.openstreetmap.org/copyright](https://www.openstreetmap.org/copyright)をご参照ください。

This project is licensed under the MIT License - see the LICENSE file for details.  
Please check [here](http://reiki.adb.shizuoka.ac.jp/act/frame/frame110000075.htm) for more information on the use of the academic emblem data.  
The campus data includes data from OpenStreetMap. Please check [here](https://www.openstreetmap.org/copyright) for more information.

## Contact

TwitterのDMあるいはE-Mailにてご連絡ください。  
Twitter: [@xR_Hamamatsu](https://twitter.com/xR_Hamamatsu)  
E-mail: xr.hamamatsu(@)gmail.com

## Author

静岡大学xRサークルはまりあ

(c) 2020,2021 Shizuoka University xR Association "Hamaria"  
(c) [OpenStreetMap](https://www.openstreetmap.org/copyright) contributors


history.txt
Ver.2.2.0
・各種ライブラリの更新と、仕様変更に伴うエラーの修正（h-takeyeahさん ありがとうございます）
・editorconfigを追加

Ver.2.1.0 (2021-11-23)
・3Dモデルを更新
・トップページに浜キャンフェスタのバナーを追加

Ver.2.0.0 (2021-09-30)
・名称を「静岡大学 バーチャル浜松キャンパス」に変更
・デザインを全面リニューアルし、モバイル端末での表示を改善
・ナイトモードを追加
・コンテンツを拡充
・プロモーションビデオを更新
・操作説明ムービーを削除
・ビデオの配信元をYouTubeに変更
・各種ライブラリを更新
・その他、プログラムの改善

Ver.1.1.2 (2021-06-27)
・一部の画像をPNGからSVGに変更
・プロモーションビデオと操作説明ムービーを更新
・トップページの画像を更新
・トップページにTwitterカードの情報を追加

Ver.1.1.1 (2021-06-19)
・「xR同好会はまりあ」→「xRサークルはまりあ」に変更
・ダイアログに表示するコンテンツの読み込み中は「読み込み中。しばらくお待ち下さい...」のメッセージを表示するようにした
・画面の右端でツールチップ内の文字列が自動改行するバグを修正
・ダイアログ表示中もツールチップが表示されるバグを修正
・README.mdとLICENSEを更新
・その他、誤字脱字の修正など

Ver.1.1 (2020-12-12)
・設定ボタン、設定画面を追加
・ダイアログを閉じても動画が再生し続けるバグを修正
・ダイアログ表示中もツールチップが出てしまうバグを修正
・ダイアログ表示中も3Dモデルがマウス操作できてしまうバグを修正
・デザインの一部変更
・プロモーションビデオ更新

Ver.1.0.4 (2020-08-22）
・「佐鳴会館」「高柳健次郎」「小ネタ」の説明文を追加
・ダイアログCSSの更新
・トップページの更新
・プロモーションビデオの修正

Ver.1.0.3 (2020-08-16)
・誤字の訂正

Ver.1.0.2 (2020-08-04)
・VRモードに操作説明の画像を追加

Ver.1.0.1 (2020-08-03)
・VRモードの戻るボタンに「Escキーでも終了できます」画像を追加

Ver.1.0 (2020-08-02)
・正式公開版に移行
・CSSを修正
・時計塔と光創起イノベーション研究拠点棟にツールチップを追加
・アクセス解析用にGoogleアナリティクスを追加

Prototype / Rev.17 (2020-08-01)
・操作説明を追加
・プロモーションビデオ更新
・スタッフクレジット更新

Prototype / Rev.16 (2020-08-01)
・モデルデータ更新
・プロモーションビデオ更新
・北会館、南会館の記述を追加

Prototype / Rev.15 (2020-07-28)
・タブレット端末用のバーチャルパッドを追加（tawatasoさん ありがとうございます）
・起動画面にプログレスバーを追加（tawatasoさん ありがとうございます）
・iOSにおけるスクロール処理を修正（tawatasoさん ありがとうございます）
・建物の表記を一部修正
・S-Port、情報学部2号館、共通講義棟、北会館、南会館の説明文と写真を追加
・スタッフクレジットを追加

Prototype / Rev.14 (2020-07-23)
・処理の最適化（tawatasoさん ありがとうございます）
・ツールチップを全建物に追加（tawatasoさん ありがとうございます）
・宣伝用のバナーを追加

Prototype / Rev.13 (2020-07-20)
・ダイアログ機能を用いたコンテンツを追加
・トップページをリニューアル
・宣伝用のバナーを追加

Prototype / Rev.12 (2020-07-12)
・トップページ「xR同好会はまりあ」リンク切れを修正
・コンテンツ表示用のウインドウを追加
・読み込み中画面に画像を追加

Prototype / Rev.11 (2020-07-11)
・（tawatasoさんによる更新）

Prototype / Rev.10 (2020-07-10)
・簡易的な読み込み中画面を追加
・トップページ「起動する」ボタンのデザインを変更

Prototype / Rev.9 (2020-07-07)
・VRButtonを追加（toyakunさん、tawatasoさん ありがとうございます）
・画面サイズ変更時に描画領域サイズが戻らなくなるバグを修正（tawatasoさん ありがとうございます）

Prototype / Rev.8 (2020-07-03)
・トップページCSSの修正
・スマートフォン／タブレット端末用のCSSを修正
・タッチパネル処理の修正（Issue #6）
・著作権表示のAdobe Illustratorファイルを追加

Prototype / Rev.7 (2020-07-02)
・VRモードを追加
・デバッグボタンを追加
・ロゴマーク、ボタンのAdobe Illustratorファイルを追加

Prototype / Rev.6 (2020-07-01)
・デザインの一部変更
・モデルを1ファイルに統合
・テクスチャのちらつきを修正（tawatasoさん ありがとうございます）
・共通講義棟にツールチップを追加
・スマートフォン／タブレット端末用のCSSを設定

Prototype / Rev.5 (2020-06-26)
・トップページと3Dキャンパスのページを分離
・情報学部2号館にマウスを重ねるとツールチップが表示される仕様に変更
・ロゴマークを変更

Prototype / Rev.4 (2020-06-19)
・情報学部2号館を独立したモデルデータに変更
・情報学部2号館をクリックするとダイアログが表示される機能の追加
・コピーライト文の追加

Prototype / Rev.3 (2020-06-14)
・デバッグ用にカメラのパラメータを表示
・カメラの描画距離を100000に変更

Prototype / Rev.2 (2020-06-11)
・ロゴを変更
・モデルデータをtoyakunさん制作の物に変更
・背景色を静大カラー（DIC 143）に変更
・オブジェクトにアンチエイリアスを設定
・スマートフォン／タブレット端末用のviewportを設定

Prototype / Rev.1 (2020-06-10)
・初版


js/VRButton.js
/**
 * @author mrdoob / http://mrdoob.com
 * @author Mugen87 / https://github.com/Mugen87
 */

var VRButton = {

	createButton: function ( renderer, options ) {

		if ( options ) {

			console.error( 'THREE.VRButton: The "options" parameter has been removed. Please set the reference space type via renderer.xr.setReferenceSpaceType() instead.' );

		}

		function showEnterVR( /*device*/ ) {

			var currentSession = null;

			function onSessionStarted( session ) {

				session.addEventListener( 'end', onSessionEnded );

				renderer.xr.setSession( session );
				button.textContent = 'EXIT VR';

				currentSession = session;

			}

			function onSessionEnded( /*event*/ ) {

				currentSession.removeEventListener( 'end', onSessionEnded );

				button.textContent = 'ENTER VR';

				currentSession = null;

			}

			//

			button.style.display = '';

			button.style.cursor = 'pointer';
			button.style.left = 'calc(50% - 50px)';
			button.style.width = '100px';

			button.textContent = 'ENTER VR';

			button.onmouseenter = function () {

				button.style.opacity = '1.0';

			};

			button.onmouseleave = function () {

				button.style.opacity = '0.5';

			};

			button.onclick = function () {

				if ( currentSession === null ) {

					// WebXR's requestReferenceSpace only works if the corresponding feature
					// was requested at session creation time. For simplicity, just ask for
					// the interesting ones as optional features, but be aware that the
					// requestReferenceSpace call will fail if it turns out to be unavailable.
					// ('local' is always available for immersive sessions and doesn't need to
					// be requested separately.)

					var sessionInit = { optionalFeatures: [ 'local-floor', 'bounded-floor' ] };
					navigator.xr.requestSession( 'immersive-vr', sessionInit ).then( onSessionStarted );

				} else {

					currentSession.end();

				}

			};

		}

		function disableButton() {

			button.style.display = '';

			button.style.cursor = 'auto';
			button.style.left = 'calc(50% - 75px)';
			button.style.width = '150px';

			button.onmouseenter = null;
			button.onmouseleave = null;

			button.onclick = null;

		}

		function showWebXRNotFound() {

			disableButton();

			button.textContent = 'VR NOT SUPPORTED';

		}

		function stylizeElement( element ) {

			element.style.position = 'absolute';
			element.style.bottom = '20px';
			element.style.padding = '12px 6px';
			element.style.border = '1px solid #fff';
			element.style.borderRadius = '4px';
			element.style.background = 'rgba(0,0,0,0.1)';
			element.style.color = '#fff';
			element.style.font = 'normal 13px sans-serif';
			element.style.textAlign = 'center';
			element.style.opacity = '0.5';
			element.style.outline = 'none';
			element.style.zIndex = '999';

		}

		if ( 'xr' in navigator ) {

			var button = document.createElement( 'button' );
			button.id = 'VRButton';
			button.style.display = 'none';

			stylizeElement( button );

			navigator.xr.isSessionSupported( 'immersive-vr' ).then( function ( supported ) {

				supported ? showEnterVR() : showWebXRNotFound();

			} );

			return button;

		} else {

			var message = document.createElement( 'a' );

			if ( window.isSecureContext === false ) {

				message.href = document.location.href.replace( /^http:/, 'https:' );
				message.innerHTML = 'WEBXR NEEDS HTTPS'; // TODO Improve message

			} else {

				message.href = 'https://immersiveweb.dev/';
				message.innerHTML = 'WEBXR NOT AVAILABLE';

			}

			message.style.left = 'calc(50% - 90px)';
			message.style.width = '180px';
			message.style.textDecoration = 'none';

			stylizeElement( message );

			return message;

		}

	}

};

export { VRButton };

js/virtualpad.js
/*
 * Developed by Shizuoka University xR Association "Hamaria"
 */

const domPad = document.querySelector('.virtualpad');
const domControl = document.querySelector('.virtualpad-control');
let vector = {x:0,y:0};
var VirtualPad = {

    init: function() {
        const initx = 5;
        const inity = 5;
        let updateEvent = new Event('virtualpadmove', {vector:{}});
        //console.log('aaa' + initx + "," + inity);

        let sx, sy = 0;
        domPad.addEventListener('touchstart', (evt) => {
            sx = evt.changedTouches[0].screenX || 0;
            sy = evt.changedTouches[0].screenY || 0;
        }, false);

        domPad.addEventListener('touchmove', (evt) => {
            const cx = evt.changedTouches[0].screenX || 0;
            const cy = evt.changedTouches[0].screenY || 0;
            let vx = cx - sx;
            let vy = cy - sy;
            const sqrt = Math.sqrt(vx*vx+vy*vy);
            vx /= sqrt;
            vy /= sqrt;
            domControl.style.top = (inity + 0 + vy * 20) + 'px';
            domControl.style.left = (initx + 0 + vx * 20) + 'px';
            vector = {x:vx, y:vy};
        }, false);

        domPad.addEventListener('touchend', (evt) => {
            sx = 0;
            sy = 0;
            domControl.style.top = inity + 'px';
            domControl.style.left = initx + 'px';
            vector = {x:0, y:0};

        }, false);
    },
    show: function() {
        domPad.style.display = 'block';
    },
    hide: function() {
        domPad.style.display = 'none';
    },
    getVector: function() {
        return vector;
    }

};

export { VirtualPad };

js/PointerLockControls.js

THREE.PointerLockControls = function ( camera, domElement ) {

	if ( domElement === undefined ) {

		console.warn( 'THREE.PointerLockControls: The second parameter "domElement" is now mandatory.' );
		domElement = document.body;

	}

	this.domElement = domElement;
	this.isLocked = false;

	// Set to constrain the pitch of the camera
	// Range is 0 to Math.PI radians
	this.minPolarAngle = 0; // radians
	this.maxPolarAngle = Math.PI; // radians

	//
	// internals
	//

	var scope = this;

	var changeEvent = { type: 'change' };
	var lockEvent = { type: 'lock' };
	var unlockEvent = { type: 'unlock' };

	var euler = new THREE.Euler( 0, 0, 0, 'YXZ' );

	var PI_2 = Math.PI / 2;

	var vec = new THREE.Vector3();
	var touchVec = new THREE.Vector3();
    //var scope = window;
	function onMouseMove( event ) {

		if(scope.isLocked === false && !isDesktop()) {
			const ts = event.changedTouches;
			if(ts.length == 1) {
				touchVec.x = ts[0].screenX;
				touchVec.y = ts[0].screenY;
			}
		}

		if ( scope.isLocked === false ) return;

		var movementX = event.movementX || event.mozMovementX || event.webkitMovementX || 0;
		var movementY = event.movementY || event.mozMovementY || event.webkitMovementY || 0;

		if(!isDesktop()) {
			const ts = event.changedTouches;
			if(ts.length == 1) {
				movementX = ts[0].screenX - touchVec.x;
				movementY = ts[0].screenY - touchVec.y;
				touchVec.x = ts[0].screenX;
				touchVec.y = ts[0].screenY;
			}
		}

		euler.setFromQuaternion( camera.quaternion );

		euler.y -= movementX * 0.002;
		euler.x -= movementY * 0.002;

		euler.x = Math.max( PI_2 - scope.maxPolarAngle, Math.min( PI_2 - scope.minPolarAngle, euler.x ) );

		camera.quaternion.setFromEuler( euler );

		scope.dispatchEvent( changeEvent );

	}

	function onPointerlockChange() {
		if ( scope.domElement.ownerDocument.pointerLockElement === scope.domElement ) {
			scope.dispatchEvent( lockEvent );
			scope.isLocked = true;
		} else {
			scope.dispatchEvent( unlockEvent );
			scope.isLocked = false;
		}
	}

	function onPointerlockError() {
		console.error( 'THREE.PointerLockControls: Unable to use Pointer Lock API' );
	}

	function onTouchDown(event) {
		const ts = event.changedTouches;
		if(ts.length == 1) {
			touchVec.x = ts[0].screenX;
			touchVec.y = ts[0].screenY;
		}
	}

	function isDesktop() {
		const ua = navigator.userAgent.toLowerCase();
		return !((('ontouchend' in document) && ua.indexOf('macintosh') > 0) || ua.indexOf('iphone') > 0 || ua.indexOf('ipod') > 0 || ua.indexOf('ipad') > 0 || ua.indexOf('android') > 0 && ua.indexOf('mobile') > 0 || ua.indexOf('android') > 0);
	}

	this.connect = function () {
		if(isDesktop()) {
			scope.domElement.ownerDocument.addEventListener( 'mousemove', onMouseMove, false );
			scope.domElement.ownerDocument.addEventListener( 'pointerlockchange', onPointerlockChange, false );
			scope.domElement.ownerDocument.addEventListener( 'pointerlockerror', onPointerlockError, false );
		} else {
			scope.domElement.addEventListener( 'touchmove', onMouseMove, false );
			scope.domElement.addEventListener( 'touchstart', onTouchDown, false );
		}
	};

	this.disconnect = function () {
		if(isDesktop()) {
			scope.domElement.ownerDocument.removeEventListener( 'mousemove', onMouseMove, false );
			scope.domElement.ownerDocument.removeEventListener( 'pointerlockchange', onPointerlockChange, false );
			scope.domElement.ownerDocument.removeEventListener( 'pointerlockerror', onPointerlockError, false );
		} else {
			scope.domElement.removeEventListener( 'touchstart', onTouchDown, false );
			scope.domElement.removeEventListener( 'touchmove', onMouseMove, false );
		}
	};

	this.dispose = function () {
		this.disconnect();
	};

	this.getObject = function () { // retaining this method for backward compatibility
		return camera;
	};

	this.getDirection = function () {
		var direction = new THREE.Vector3( 0, 0, - 1 );
		return function ( v ) {
			return v.copy( direction ).applyQuaternion( camera.quaternion );
		};
	}();

	this.moveForward = function ( distance ) {

		// move forward parallel to the xz-plane
		// assumes camera.up is y-up

		vec.setFromMatrixColumn( camera.matrix, 0 );
		vec.crossVectors( camera.up, vec );
		camera.position.addScaledVector( vec, distance );
	};

	this.moveRight = function ( distance ) {
		vec.setFromMatrixColumn( camera.matrix, 0 );
		camera.position.addScaledVector( vec, distance );
	};

	this.lock = function () {
		if(isDesktop()) {
			this.domElement.requestPointerLock();
		} else {
			scope.dispatchEvent( lockEvent );
			scope.isLocked = true;
		}
	};

	this.unlock = function () {
		if(isDesktop()) {
			scope.domElement.ownerDocument.exitPointerLock();
		} else {
			scope.dispatchEvent( unlockEvent );
			scope.isLocked = false;
		}
	};

	this.desktopMode = isDesktop();

	this.connect();

};

THREE.PointerLockControls.prototype = Object.create( THREE.EventDispatcher.prototype );
THREE.PointerLockControls.prototype.constructor = THREE.PointerLockControls;

//export { PointerLockControls };

js/WebVR.js
/**
 * @author mrdoob / http://mrdoob.com
 * @author Mugen87 / https://github.com/Mugen87
 *
 * Based on @tojiro's vr-samples-utils.js
 */
var polyfill = new WebXRPolyfill();
var supportOldAPI = false;

var WEBVR = {

	createButton: function ( renderer ) {

		var eventEnterVR = new CustomEvent('enter_vr', {  });
		var eventExitVR = new CustomEvent('exit_vr', {  });

		function showEnterVR( device ) {
			button.style.display = '';
			button.style.cursor = 'pointer';
			button.style.left = 'calc(50% - 50px)';
			button.style.width = '100px';

			button.textContent = 'ENTER VR';
			button.onmouseenter = function () { button.style.opacity = '1.0'; };
			button.onmouseleave = function () { button.style.opacity = '0.5'; };

			button.onclick = function () {
				device.isPresenting ? device.exitPresent() : device.requestPresent( [ { source: renderer.domElement } ] );
				return true;
			};

			//renderer.xr.setDevice( device );

		}

		function showEnterXR( device ) {
			var currentSession = null;
			function onSessionStarted( session ) {

				session.addEventListener( 'end', onSessionEnded );

				renderer.xr.setSession( session );
				button.textContent = 'EXIT XR';

				currentSession = session;
				window.dispatchEvent(eventEnterVR);

			}

			function onSessionEnded( event ) {
				currentSession.removeEventListener( 'end', onSessionEnded );

				renderer.xr.setSession( null );
				button.textContent = 'ENTER XR';

				currentSession = null;
				window.dispatchEvent(eventExitVR);
			}

			//

			button.style.display = '';

			button.style.cursor = 'pointer';
			button.style.left = 'calc(50% - 50px)';
			button.style.width = '100px';

			button.textContent = 'ENTER XR';

			button.onmouseenter = function () { button.style.opacity = '1.0'; };
			button.onmouseleave = function () { button.style.opacity = '0.5'; };

			button.onclick = function () {
				if ( currentSession === null ) {
					device.requestSession( { exclusive: true } ).then( onSessionStarted );
				} else {
					currentSession.end();
				}
			};

			//renderer.xr.setDevice( device );

		}

		function showVRNotFound() {

			button.style.display = '';

			button.style.cursor = 'auto';
			button.style.left = 'calc(50% - 75px)';
			button.style.width = '150px';

			button.textContent = 'VR NOT FOUND';

			button.onmouseenter = null;
			button.onmouseleave = null;

			button.onclick = null;

			//renderer.xr.setDevice( null );

		}

		function stylizeElement( element ) {
			element.style.position = 'absolute';
			element.style.bottom = '20px';
			element.style.padding = '12px 6px';
			element.style.border = '1px solid #fff';
			element.style.borderRadius = '4px';
			element.style.background = 'transparent';
			element.style.color = '#fff';
			element.style.font = 'normal 13px sans-serif';
			element.style.textAlign = 'center';
			element.style.opacity = '0.5';
			element.style.outline = 'none';
			element.style.zIndex = '999';

		}

		var isWebXR = false;
/*
		if ( 'xr' in navigator ) {
			isWebXR = true;
			var button = document.createElement( 'button' );
			button.style.display = 'none';
			stylizeElement( button );
			navigator.xr.requestDevice().then( function ( device ) {
				device.supportsSession( { exclusive: true } ).then( function () {
					showEnterXR( device );
				} ).catch( showVRNotFound );
			} ).catch( showVRNotFound );
			return button;
		} else
		*/
		if ( 'getVRDisplays' in navigator ) {

			var button = document.createElement( 'button' );
			button.style.display = 'none';

			stylizeElement( button );

			window.addEventListener( 'vrdisplayconnect', function ( event ) {
				showEnterVR( event.display );
			}, false );

			window.addEventListener( 'vrdisplaydisconnect', function ( event ) {
				showVRNotFound();
			}, false );

			window.addEventListener( 'vrdisplaypresentchange', function ( event ) {
				button.textContent = event.display.isPresenting ? 'EXIT VR' : 'ENTER VR';
			}, false );

			window.addEventListener( 'vrdisplayactivate', function ( event ) {

				event.display.requestPresent( [ { source: renderer.domElement } ] );

			}, false );

			navigator.getVRDisplays()
				.then( function ( displays ) {
					if ( displays.length > 0 ) {
						showEnterVR( displays[ 0 ] );
					} else {
						showVRNotFound();
					}
				} );
			return button;
		} else {
			var message = document.createElement( 'a' );
			message.href = 'https://webvr.info';
			message.innerHTML = 'WEBVR NOT SUPPORTED';
			message.style.left = 'calc(50% - 90px)';
			message.style.width = '180px';
			message.style.textDecoration = 'none';
			stylizeElement( message );
			return message;
		}
	},

	// DEPRECATED
	checkAvailability: function () {
		console.warn( 'WEBVR.checkAvailability has been deprecated.' );
		return new Promise( function () {} );
	},

	getMessageContainer: function () {
		console.warn( 'WEBVR.getMessageContainer has been deprecated.' );
		return document.createElement( 'div' );
	},

	getButton: function () {
		console.warn( 'WEBVR.getButton has been deprecated.' );
		return document.createElement( 'div' );
	},

	getVRDisplay: function () {
		console.warn( 'WEBVR.getVRDisplay has been deprecated.' );
	}

};


var WEBXR = {
	createButton: function ( renderer, options ) {
		if ( options ) {
			console.error( 'THREE.VRButton: The "options" parameter has been removed. Please set the reference space type via renderer.xr.setReferenceSpaceType() instead.' );
		}
		
		var eventEnterVR = new CustomEvent('enter_vr', {  });
		var eventExitVR = new CustomEvent('exit_vr', {  });

		function showEnterVR( /*device*/ ) {
			var currentSession = null;
			function onSessionStarted( session ) {
				session.addEventListener( 'end', onSessionEnded );
				renderer.xr.setSession( session );
				button.textContent = 'EXIT VR';
				currentSession = session;
				window.dispatchEvent(eventEnterVR);
			}

			function onSessionEnded( /*event*/ ) {
				currentSession.removeEventListener( 'end', onSessionEnded );
				button.textContent = 'ENTER VR';
				currentSession = null;
				window.dispatchEvent(eventExitVR);
			}

			button.style.display = '';
			button.style.cursor = 'pointer';
			button.style.left = 'calc(50% - 50px)';
			button.style.width = '100px';
			button.textContent = 'ENTER VR';

			button.onmouseenter = function () {
				button.style.opacity = '1.0';
			};

			button.onmouseleave = function () {
				button.style.opacity = '0.5';
			};

			button.onclick = function () {
				if ( currentSession === null ) {

					// WebXR's requestReferenceSpace only works if the corresponding feature
					// was requested at session creation time. For simplicity, just ask for
					// the interesting ones as optional features, but be aware that the
					// requestReferenceSpace call will fail if it turns out to be unavailable.
					// ('local' is always available for immersive sessions and doesn't need to
					// be requested separately.)

					var sessionInit = { optionalFeatures: [ 'local-floor', 'bounded-floor' ] };
					navigator.xr.requestSession( 'immersive-vr', sessionInit ).then( onSessionStarted );

				} else {
					currentSession.end();
				}
				return true;
			};
		}
		function disableButton() {
			button.style.display = '';
			button.style.cursor = 'auto';
			button.style.left = 'calc(50% - 75px)';
			button.style.width = '150px';
			button.onmouseenter = null;
			button.onmouseleave = null;
			button.onclick = null;
		}

		function showWebXRNotFound() {
			disableButton();
			button.textContent = 'VR NOT SUPPORTED';
		}

		function stylizeElement( element ) {
			element.style.position = 'absolute';
			element.style.bottom = '20px';
			element.style.padding = '12px 6px';
			element.style.border = '1px solid #fff';
			element.style.borderRadius = '4px';
			element.style.background = 'rgba(0,0,0,0.1)';
			element.style.color = '#fff';
			element.style.font = 'normal 13px sans-serif';
			element.style.textAlign = 'center';
			element.style.opacity = '0.5';
			element.style.outline = 'none';
			element.style.zIndex = '999';
		}

		if ( 'xr' in navigator || navigator.xr !== undefined) {
			var button = document.createElement( 'button' );
			button.id = 'VRButton';
			button.style.display = 'none';
			stylizeElement( button );
			if(navigator.xr.isSessionSupported) {
				navigator.xr.isSessionSupported( 'immersive-vr' ).then( function ( supported ) {
					supported ? showEnterVR() : showWebXRNotFound();
				} );
			} else if(navigator.xr.supportsSession) {
				navigator.xr.supportsSession( 'immersive-vr' ).then( function () {
					showEnterVR();
				} );
			}
			return button;
		} else {
			var message = document.createElement( 'a' );
			if ( window.isSecureContext === false ) {
				message.href = document.location.href.replace( /^http:/, 'https:' );
				message.innerHTML = 'WEBXR NEEDS HTTPS'; // TODO Improve message
			} else {
				message.href = 'https://immersiveweb.dev/';
				message.innerHTML = 'WEBXR NOT AVAILABLE';
			}


			message.style.left = 'calc(50% - 90px)';
			message.style.width = '180px';
			message.style.textDecoration = 'none';
			stylizeElement( message );
			return message;
		}
	}
};


var VRButton = {
	
	createButton : function(renderer, options) {

		//console.log('getVRDisplays' in navigator);
		if ( supportOldAPI && 'getVRDisplays' in navigator && !('xr' in navigator)) {
			return WEBVR.createButton(renderer);
		}

		return WEBXR.createButton(renderer);

	},

	enableVR : function() {
		return 'xr' in navigator || (supportOldAPI && 'getVRDisplays' in navigator);
	}
};


export { VRButton };
